astropy/modeling/astro_sherpa.py

```diff
@@ -0,0 +1,433 @@
+    If deconstruct is true then the submodels are converted then recombined.
+
+    Parameters
+    ----------
+        model : `~astropy.modeling.FittableModel`
+            the model you wish to convert
+        diconstruct: bool
+            if True each submodel is converted then re-assembled based on the
+            model._format_expression()
+    """
+
+    def _convert_to_sherpa_model(model):
+        '''
+        Converts the model using sherpa's usermodel suppling the parameter detail to sherpa
+        then using a decorator to allow the call method to act like the calc method
+        '''
+        def _calc2call(func):
+            '''This decorator makes call and calc work together.'''
+            def _converter(inp, *x):
+                if func.n_inputs == 1:
+                    # startp = func._parameters[:]
+                    # func._parameters = np.array(inp)
+                    # retvals = func.__call__(*x)
+                    retvals = func.evaluate(x[0], *inp)
+                    # func._parameters = startp
+                else:
+                    retvals = func.evaluate(x[0], x[1], *inp)  # I need to test the 2d case!
+                return retvals
+            return _converter
+
+        pars = []
+        linkedpars = []
+        for pname in model.param_names:
+            param = model.__getattribute__(pname)
+            vals = [param.name, param.value, param.min, param.max, param.min,
+                    param.max, None, param.fixed, False]
+            attrnames = ["name", "val", "min", "max", "hard_min", "hard_max",
+                         "units", "frozen", "alwaysfrozen"]
+            if model.name is None:
+                model._name = ""
+
+            pars.append(Parameter(modelname="wrap_" + model.name, **dict([(atr, val) for atr, val in zip(attrnames, vals) if val is not None])))
+            if param.tied is not False:
+                linkedpars.append(pname)
+
+        smodel = UserModel(model.name, pars)
+        smodel.calc = _calc2call(model)
+
+        for pname in linkedpars:
+            param = model.__getattribute__(pname)
+            sparam = smodel.__getattribute__(pname)
+            sparam.link = param.tied(smodel)
+
+        return smodel
+
+    if deconstruct:
+        models = [_convert_to_sherpa_model(m) for m in model.submodels]
+        expression = model._format_expression()
+
+        if isinstance(models, list) or isinstance(models, tuple):
+            models = {str(num): mod for num, mod in enumerate(models)}
+
+            expression_list = _expression_to_list(expression)
+            return [_parse_expression(elist, models)[0] for elist in expression_list]
+        else:
+            expression_list = _expression_to_list(expression)
+            return _parse_expression(expression_list, models)[0]
+    else:
+        return _convert_to_sherpa_model(model)
+
+
+def _make_dataset(n_dim, x, y, z=None, xerr=None, yerr=None, zerr=None, n=0):
+    '''
```

Comment:
I believe PEP8 mandates tripe double-quotes for docstrings. Nothing serious, of course. Also, some docstrings seem to be out of sync with the code. Again, I wouldn't worry too much at this stage.